design/src/rfcs/rfc0012_unsynchronized_versioning.md

```diff
@@ -0,0 +1,207 @@
+
+[aws-smithy-types]
+version = "0.50.1"
+
+[aws-config]
+version = "0.40.0"
+
+[aws-sdk-s3]
+version = "0.89.0"
+model_hash = "<hash>"
+
+# ...
+```
+
+The auto-sync tool is responsible for maintaining this file. When it generates a new SDK, it will take
+the version numbers from runtime crates directly, and it will use the rules from the next section to determine
+the version numbers for the generated crates.
+
+### Versioning for Code Generated (SDK Service) Crates
+
+Code generated crates will have their `minor` version bumped when the version of [smithy-rs] used to generate
+them changes, or when model updates with API changes are made. Three pieces of information are required to
+handle this process: the previously released version number, the [smithy-rs] version used to generate the code,
+and the level of model updates being applied. For this last one, if there are multiple model updates that
+affect only documentation, but then one model update that affects an API, then as a whole they will be
+considered as affecting an API and require a `minor` version bump.
+
+The previously released version number will be retrieved from crates.io using its API. The [smithy-rs] version
+used during code generation will become a build artifact that is saved to `versions.toml` in [aws-sdk-rust].
+During phase 1, the tooling required to know if a model is a documentation-only change will not be available,
+so all model changes will result in a `minor` version bump during this phase.
+
+Overall, determining a generated crate's version number looks as follows:
+
+![Phase 1: How to version a generated crate](rfc0012_unsynchronized_versioning/phase1_generated_crate_version.svg)
+
+- __A: smithy-rs changed?__: Compare the `smithy_rs_version` in the previous `versions.toml` with the
+  next `versions.toml` file, and if the values are different, consider [smithy-rs] to have changed.
+- __B: model changed?__: Similarly, compare the `model_hash` for the crate in `versions.toml`.
+
+### Versioning for Runtime Crates
+
+The old scheme of all runtime crates in [smithy-rs] having a fake `0.0.0-smithy-rs-head` version number with
+a build step to replace those with a consistent `major.minor` will be removed. These runtime crates will begin
+having their actual next version number in the Cargo.toml file in smithy-rs.
+
+This introduces a new problem where a developer can forget to bump a runtime crate version, so a method of
+process enforcement needs to be introduced. This will be done through CI when merging into `smithy-rs/main`
+and repeated when merging into `aws-sdk-rust/main`.
+
+The following checks need to be run for runtime crates:
+
+![Phase 1: How to validate a runtime version bump](rfc0012_unsynchronized_versioning/phase1_runtime_crate_version_checks.svg)
+
+- __A: Crate has changed?__ The crate's source files and manifest will be hashed for the previous version
+  and the next version. If these hashes match, then the crate is considered unchanged.
```

Comment:
Cargo has something like this for incremental compilation right? It'd be kinda neat if we could use that functionality to do this.

Reply:
It would be neat, but I think it would require building a tool that utilizes the compiler internals, and introduce the maintenance burden that comes with that kind of tooling.